### DISCUSSION TITLE
MDS in readonly mode:  mds.0.journaler.pq(ro) _finish_read got less than expected (1959859)

### DISCUSSION THREAD

**Discussion by osaffer** [rook/rook#9345] [General]
Hello,

After a problem on our physical storage, we started to get problem on our rook-ceph infrastructure.
MountVolume.MountDevice failed for volume "pvc-63538b29-0a0f-4c16-8b72-2bbc7f994082" : rpc error: code = Internal desc = rados: ret=-30, Read-only file system: "error in setxattr"

Then I check status of the ceph cluster, indeed : 
 ceph health detail
HEALTH_WARN 1 MDSs are read only; mons aa,ab are low on available space; 1 pool(s) have no replicas configured; 44 daemons have recently crashed
[WRN] MDS_READ_ONLY: 1 MDSs are read only
    mds.postgresfs-b(mds.0): MDS in read-only mode

I verfied logs in mds, there were some duplicated inodes.
IT has been fixed with :   
cephfs-data-scan scan_links

2021-12-07T20:43:47.254+0000 7f78445a6440 -1 datascan.scan_links: Remove duplicated 
 
I also did :  

ceph tell mds.postgresfs-b scrub start / force recursive repair

cephfs-journal-tool event recover_dentries summary
cephfs-table-tool all reset session
cephfs-journal-tool journal reset
cephfs-data-scan init
cephfs-data-scan scan_extents data
cephfs-data-scan scan_inodes data

Now I have the following errors : 

debug 2021-12-08T07:44:12.640+0000 7f5e0ab2b780  0 set uid:gid to 167:167 (ceph:ceph)
debug 2021-12-08T07:44:12.640+0000 7f5e0ab2b780  0 ceph version 16.2.6 (ee28fb57e47e9f88813e24bbf4c14496ca299d31) pacific (stable), process ceph-mds, pid 1
debug 2021-12-08T07:44:12.640+0000 7f5e0ab2b780  1 main not setting numa affinity
debug 2021-12-08T07:44:12.640+0000 7f5e0ab2b780  0 pidfile_write: ignore empty --pid-file
starting mds.postgresfs-b at 
debug 2021-12-08T07:44:12.652+0000 7f5df8c8e700  1 mds.postgresfs-b Updating MDS map to version 29296 from mon.0
debug 2021-12-08T07:44:13.240+0000 7f5df8c8e700  1 mds.postgresfs-b Updating MDS map to version 29297 from mon.0
debug 2021-12-08T07:44:13.240+0000 7f5df8c8e700  1 mds.postgresfs-b Monitors have assigned me to become a standby.
debug 2021-12-08T07:44:13.264+0000 7f5df8c8e700  1 mds.postgresfs-b Updating MDS map to version 29298 from mon.0
debug 2021-12-08T07:44:13.264+0000 7f5df8c8e700  1 mds.0.29298 handle_mds_map i am now mds.0.29298
debug 2021-12-08T07:44:13.264+0000 7f5df8c8e700  1 mds.0.29298 handle_mds_map state change up:boot --> up:replay
debug 2021-12-08T07:44:13.264+0000 7f5df8c8e700  1 mds.0.29298 replay_start
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700  0 mds.0.journaler.pq(ro) _finish_read got less than expected (1959859)
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700  0 mds.0.journaler.pq(ro) _finish_read got error -2
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700 -1 mds.0.purge_queue _recover: Error -2 recovering write_pos
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700  1 mds.0.purge_queue _go_readonly: going readonly because internal IO failed: No such file or directory
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700 -1 mds.0.29298 unhandled write error (2) No such file or directory, force readonly...
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700  1 mds.0.cache force file system read-only
debug 2021-12-08T07:44:13.272+0000 7f5df3c84700  0 log_channel(cluster) log [WRN] : force file system read-only
debug 2021-12-08T07:44:13.280+0000 7f5df2c82700  0 mds.0.cache creating system inode with ino:0x100
debug 2021-12-08T07:44:13.280+0000 7f5df2c82700  0 mds.0.cache creating system inode with ino:0x1
debug 2021-12-08T07:44:13.300+0000 7f5df147f700  1 mds.0.journal EResetJournal
debug 2021-12-08T07:44:13.300+0000 7f5df147f700  1 mds.0.sessionmap wipe start
debug 2021-12-08T07:44:13.300+0000 7f5df147f700  1 mds.0.sessionmap wipe result
debug 2021-12-08T07:44:13.300+0000 7f5df147f700  1 mds.0.sessionmap wipe done
debug 2021-12-08T07:44:13.304+0000 7f5df147f700  0 mds.0.29298 boot error forcing transition to read-only; MDS will try to continue
debug 2021-12-08T07:44:13.304+0000 7f5df147f700  1 mds.0.29298 Finished replaying journal
debug 2021-12-08T07:44:13.304+0000 7f5df147f700  1 mds.0.29298 making mds journal writeable

I am trying this procedure :   

[root@rook-ceph-tools-96c99fbf-zd2k7 /]# cephfs-journal-tool --rank=1:0 event recover_dentries summary
Events by type:
  RESETJOURNAL: 1
  SESSION: 677
  UPDATE: 11
Errors: 0
[root@rook-ceph-tools-96c99fbf-zd2k7 /]# cephfs-journal-tool --rank=1:0 journal reset
old journal was 1251664199680~212391
new journal start will be 1251668393984 (3981913 bytes past old end)
writing journal head
writing EResetJournal entry
done
[root@rook-ceph-tools-96c99fbf-zd2k7 /]# cephfs-table-tool all reset session
{
    "0": {
        "data": {},
        "result": 0
    }
}

[root@rook-ceph-tools-96c99fbf-zd2k7 /]#  ceph fs reset postgresfs --yes-i-really-mean-it
Error EINVAL: all MDS daemons must be inactive before resetting filesystem: set the cluster_down flag and use `ceph mds fail` to make this so
[root@rook-ceph-tools-96c99fbf-zd2k7 /]# ceph fs set postgresfs cluster_down true
postgresfs marked not joinable; MDS cannot join as newly active. WARNING: cluster_down flag is deprecated and will be removed in a future version. Please use "joinable".
[root@rook-ceph-tools-96c99fbf-zd2k7 /]# ceph mds fail 0
failed mds gid 36739248
[root@rook-ceph-tools-96c99fbf-zd2k7 /]#  ceph fs reset postgresfs --yes-i-really-mean-it

Still running ...

Is anybody has experienced this kind of problem ?

Am I doing well ?

Thanks a lot for your help

**Comment by BlaineEXE** [comment]
I think recovering the Ceph cluster back to a healthy state is the first thing to address. While Ceph is unstable, I'm not sure filesystem recovery efforts will be able to complete.

> 44 daemons have recently crashed

What does `ceph crash ls` show? Do you still have crashes happening regularly? I think it would be best to first address crashing daemons, unless all of those crashes are mds...

**Reply by osaffer** [reply]
Hi,

The faster way to recover the cluster was to restore a backup of the VM.

I have been able to get the cluster up & running with many pg inconsistents etc ...

Finally the health is still in Error, because of a PG that I cannot repair.

 ceph health detail
HEALTH_ERR 1 MDSs report slow requests; mons aa,w are low on available space; 1 scrub errors; Possible data damage: 1 pg inconsistent; 1 pool(s) have no replicas configured
[WRN] MDS_SLOW_REQUEST: 1 MDSs report slow requests
    mds.postgresfs-b(mds.0): 1289 slow requests are blocked > 30 secs
[WRN] MON_DISK_LOW: mons aa,w are low on available space
    mon.w has 26% avail
    mon.aa has 27% avail
[ERR] OSD_SCRUB_ERRORS: 1 scrub errors
[ERR] PG_DAMAGED: Possible data damage: 1 pg inconsistent
    pg 8.48 is active+clean+inconsistent+failed_repair, acting [14]
[WRN] POOL_NO_REDUNDANCY: 1 pool(s) have no replicas configured
    pool 'replicapool' has no replicas configured



2021-12-16T09:32:13.275530+0000 osd.14 [ERR] repair 8.48 8:130b5d36:::rbd_data.cc0f3c438228bf.0000000000000608:6e : is an unexpected clone
2021-12-16T09:32:13.275575+0000 osd.14 [ERR] 8.48 repair 1 errors, 0 fixed

I found different topics explaining how to fix with ceph-objectstore-tool, but I have not been able to repair.

If you have experience about it how to fix, how to get a good procedure.

Thank you

**Reply by osaffer** [reply]
Otherwise : 

ceph crash ls
ID                                                                ENTITY            NEW  
2021-02-10T19:37:20.223520Z_c0d7b87c-3e9a-4f10-835b-9b5aaee6972c  osd.0              *   
2021-02-10T19:42:46.461661Z_01ae3ad0-555a-406b-b432-134fb8f9e1a1  osd.0              *   
2021-02-10T20:08:03.836692Z_d9c88bb5-005e-4cf4-bae6-94147cb52a24  osd.16             *   
2021-02-10T20:09:29.422086Z_e03e8ba7-7090-4722-a0cf-7193b03d7230  osd.17             *   
2021-02-10T20:10:58.030168Z_9bc0d5d4-06d2-45a9-892d-d9126ea394e1  osd.15             *   
2021-02-10T20:14:23.620823Z_80cabed6-94b6-410c-ae83-fa757b2b888c  osd.15             *   
2021-02-10T20:14:23.620875Z_7487c471-00c8-401a-81e5-b9b2623420ef  osd.16             *   
2021-02-10T20:42:53.197000Z_e641a505-9f1b-42b5-8345-d6c8451f8e1d  osd.4              *   
2021-02-10T21:05:51.445040Z_56f5d00c-9f69-4f01-9831-22ee81d40f36  osd.2              *   
2021-02-10T21:05:51.452008Z_ac21e988-c287-4e68-afd8-b9b3904852eb  osd.5              *   
2021-02-10T21:05:51.452819Z_b99d7701-d1a9-4613-b235-73bab39bcdfa  osd.5              *   
2021-02-10T21:05:51.462021Z_02317da5-7916-4c39-958d-83c5e6898873  osd.2              *   
2021-05-25T00:19:37.132319Z_73b02d94-0638-4b6a-84f6-b0856ee33440  osd.17             *   
2021-05-25T00:19:37.239483Z_698a5fc5-f802-4b7d-83fe-b4564d9a2bd6  osd.15             *   
2021-05-25T00:19:37.709547Z_c67fe2fe-018b-46ea-bc35-c45db2001757  osd.17             *   
2021-05-25T00:19:37.713320Z_bd19476d-7d6c-4297-a67e-c48384047133  osd.15             *   
2021-09-30T18:45:46.515523Z_5f0d3083-90e3-4bdb-b5f3-0172aa565829  mgr.a              *   
2021-10-01T08:52:38.448463Z_f1f1ef7a-5886-470a-9db9-349636901fc7  mds.postgresfs-b   *   
2021-10-01T08:53:42.814036Z_2534e8d4-a414-45a2-a1eb-75e03bac1077  mds.postgresfs-b   *   
2021-10-11T08:40:33.035681Z_aff3e6b4-6a0b-48ea-89ed-d9014344714c  mds.postgresfs-b   *
